### ISSUE TITLE
Text selection is not remembered when scrolling

### ISSUE THREAD

**Issue by kekkc** [mozilla/pdf.js#18221]
Steps to reproduce the problem:
1. Select some text (e.g. 

> Dynamic languages such as JavaScript are more difﬁcult to com-

2. Scroll to the last page of the pdf and then scroll to the top again

What is the expected behavior? 
Selection is still marked in blue (so that I can see where I stopped reading)

What went wrong? 
Selection is gone, no text is selected any more

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):
https://mozilla.github.io/pdf.js/web/viewer.html

**Comment by Adarsh01208** [comment]
I have also faced this issue

**Comment by Legus-Yeung** [comment]
When I replicate this issue, I highlighted the text on the first page (1) and the words will stay highlighted as long as I don't scroll past page (8). If I highlight on the first page and scroll to page 9 and onwards, the highlighted text stays.

[](https://mozilla.github.io/pdf.js/web/viewer.html)
 
This makes me think that pdf viewer will "forget/unload" the highlighted text as you render more pages in order to preserve memory. So it might not be a static occurrence from scrolling 8 pages away, if the pages in between have a lot of content like HD pictures, the highlighted text might disappear sooner than 8 pages.

Suggested solution: Create a variable or a sidebar with a functionality to store the highlighted text, this may be very useful for people who are doing research on a pdf.

**Comment by borontion** [comment]
There is a buffer for page views and views exceed the buffer size will be destoryed

https://github.com/mozilla/pdf.js/blob/7a8aceef20925dc082a8f1f863134958ceb46015/web/pdf_viewer.js#L147-L149

The buffer size will be 10 for this document

https://github.com/mozilla/pdf.js/blob/7a8aceef20925dc082a8f1f863134958ceb46015/web/pdf_viewer.js#L1634

**Comment by kekkc** [comment]
Nice catch, but there seems to be no ease way to disable this buffer.

BTW: just tested with Chrome and it preserves the selection and also the pages. Even with a 500 pages PDF Chrome seems to using no cache or buffer at all (a usual website with 500 pages also has no buffer).

**Comment by QuantumAbraham** [comment]
> BTW: just tested with Chrome and it preserves the selection and also the pages. Even with a 500 pages PDF Chrome seems to using no cache or buffer at all (a usual website with 500 pages also has no buffer).

Yes the issue is not present in chrome but they do use buffers, its only that the text selection is preserved

**Comment by kekkc** [comment]
> Yes the issue is not present in chrome but they do use buffers, its only that the text selection is preserved
There's no delay in Chrome for 500 pages, so I assumed that there's also no buffer (in contrast there's always the loading icon in FF and I cannot see the content while fast scrolling through the 500 pages in 2 seconds).

**Comment by kekkc** [comment]
Update: seems Chrome's PDFium is using caching (https://groups.google.com/g/pdfium/c/_zi2mMDiyjo), but only within C++. Javascript has it's own garbage collection within FF already. I really see no need here to stick to this 15+ years old, performance impacting buffer.

**Comment by calixteman** [comment]
> Update: seems Chrome's PDFium is using caching (https://groups.google.com/g/pdfium/c/_zi2mMDiyjo), but only within C++. Javascript has it's own garbage collection within FF already. I really see no need here to stick to this 15+ years old, performance impacting buffer.

@kekkc, could you elaborate about performance impacting buffer ?

Why removing the `PDFPageViewBuffer` would fix the text selection issue ?

**Comment by Snuffleupagus** [comment]
Please note that `PDFPageViewBuffer` exists for good reason, and it cannot just be removed since that'd cause significant problems in general:
 - It would affect memory usage very badly in long PDF documents, i.e. ones consisting of hundreds or thousands of pages.
 - It would affect overall performance *really* badly in long PDF documents, since it'd allow an arbitrary number of pages to compete for parsing and rendering resources which could easily cause the viewer to become unusable.

**Comment by kekkc** [comment]
@calixteman 
Scrolling performance is bad. If I'm at page 500 and scroll up fast to the top, I only see the loading icon on the pages. E.g. for pdfium I still see the content and images while scrolling and can stop eventually if I found what I was looking for. If there would no buffer, no pages would be deleted in RAM and also no selection would be deleted.

@Snuffleupagus 
> It would affect memory usage very badly in long PDF documents, i.e. ones consisting of hundreds or thousands of pages.

would love to test this and to set the buffer to a trillion pages, but unfortunately build env doc is not listing all dependencies like python that I cannot install directly here. Would awesome if someone would have a https://mozilla.github.io/pdf.js/web/viewer.html link to test this is real world.

>It would affect overall performance really badly, particularly in long PDF documents, since it'd allow an arbitrary number of pages to compete for parsing and rendering resources which could easily cause the viewer to become unusable.

I don't see why this is different from any other usual webpage in Firefox. If a website has 1 trillion pages and the user has thousand tabs open, FF might become unresponsive. But FF wouldn't just start to "flush" the first pages of the active website in use just because it is big. And even if it would, it should be the decision up to the browser how to distribute available parsing and render resources for the active website and across tabs and CPU threads and garbage collection.

**Comment by kekkc** [comment]
> That said in term of security, pdf.js is a way better than pdfium because it's written in js when pdfium is written in C++.

Indeed that's the main reason why I replaced any "native" local pdf reader with pdf.js in FF (possible after forms support), which is the only one that uses the sharp sub-pixel rendering from FF on HiDPI ;) Subjectively I feel no real performance penalty for rendering the first time using JS compared to C++ / Rust.

> I'm not sure to understand exactly what you mean by "other usual webpage"... 

Yes, for infinite scrolling a Webdev is responsible that the infinity will always work. But PDFs & static websites are not infinite and the end & size is known. Even for the current recordholder and longest static website worldwide, no browser would simply delete the first pages when you scrolled to the end, there's also no performance penalty (any JS operation can be executed min. 15.000 times every millisecond, even more on current ARM chips!):
http://www.recordholders.org/de/laengste/lang-a.htm

However, I did see unresponsiveness with PDFs that only have a few pages, but hundreds of Mb images. Browser as well as pdf.js were unresponsive, but of course the cache buffer doesn't prevent that and even "native" PDF readers became unresponsive there.

>The solution for the original problem (the selection one) is likely something around caching the selection when the page is destroyed and restoring it when the page is shown again.

Or increasing the buffer to 10.000 pages ;) Seriously, are there tests available that showed that this buffer is needed or why this page size is reasonable? Build environment is failing for me, but if someone could provide a pdf.js/web/viewer.html link with adjusted cache size, I could do some testing with world's most famous books for CPU time, threads & RAM:
 - https://www.forgottenbooks.com/es/download/TheIllustratedBibleTreasury_10099915.pdf (347 image pages, scrolling performance horrible in native readers and pdf.js, but native readers seem to read more pages ahead in parallel threads)
 - http://triggs.djvu.org/djvu-editions.com/BIBLES/DRV/Download.pdf (1602 text pages)

If all fails of course and there is a real world reason for the buffer, caching the selection would be cool, since I feel it's the only disadvantage left compared to native PDF readers.

**Comment by kekkc** [comment]
Was able to test in a separate VM:
buffer is still used for loading new pages, but "destroying" the old pages is disabled. Performance is similar as with native PDF readers & scrolling to the top shows the actual content. No problem with the big PDFs above (if some retests, min. 2 pages should be shown on the screen while scrolling). 

The selection is of course preserved. Just removed the following lines:
https://github.com/mozilla/pdf.js/blob/7a8aceef20925dc082a8f1f863134958ceb46015/web/pdf_viewer.js#L147-L149
https://github.com/mozilla/pdf.js/blob/7a8aceef20925dc082a8f1f863134958ceb46015/web/pdf_viewer.js#L177-L179
https://github.com/mozilla/pdf.js/blob/7a8aceef20925dc082a8f1f863134958ceb46015/web/pdf_viewer.js#L190-L195

**Comment by kekkc** [comment]
Did a PR, bottom line is IMO: no need to cache selections, just pages with selections shouldn't be destroyed. Or easier no page should be destroyed if there's no specific reason, or there could be setting in FF.

**Comment by QuantumAbraham** [comment]
@calixteman 
I agree with you on the solution you suggested for this issue, I've been thinking about it myself. Either we: 
1. Persist Text Selections: To cache text selections separately from the page buffer. This would involve saving the selection data when a page is destroyed and restoring it when the page is reloaded.
OR
2. Page Destruction Strategy: We modify the logic to prevent pages with selections from being destroyed.
Implement a setting to adjust how pdf.js handles page buffering and destruction.

@Snuffleupagus and @kekkc  what do y'all think about this?
